fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/TypeCoercion.java

```diff
@@ -0,0 +1,173 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.expression.rewrite.rules;
+
+import org.apache.doris.nereids.annotation.Developing;
+import org.apache.doris.nereids.rules.expression.rewrite.AbstractExpressionRewriteRule;
+import org.apache.doris.nereids.rules.expression.rewrite.ExpressionRewriteContext;
+import org.apache.doris.nereids.trees.expressions.BinaryOperator;
+import org.apache.doris.nereids.trees.expressions.CaseWhen;
+import org.apache.doris.nereids.trees.expressions.Cast;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.InPredicate;
+import org.apache.doris.nereids.trees.expressions.typecoercion.ImplicitCastInputTypes;
+import org.apache.doris.nereids.types.DataType;
+import org.apache.doris.nereids.types.coercion.AbstractDataType;
+import org.apache.doris.nereids.util.TypeCoercionUtils;
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * a rule to add implicit cast for expressions.
+ */
+@Developing
+public class TypeCoercion extends AbstractExpressionRewriteRule {
+
+    public static final TypeCoercion INSTANCE = new TypeCoercion();
+
+    @Override
+    public Expression rewrite(Expression expr, ExpressionRewriteContext ctx) {
+        if (expr instanceof ImplicitCastInputTypes) {
+            return visitImplicitCastInputTypes(expr, ctx);
+        } else {
+            return super.rewrite(expr, ctx);
+        }
+    }
+
+    // TODO: add other expression visitor function to do type coercion if necessary.
+
+    @Override
+    public Expression visitBinaryOperator(BinaryOperator binaryOperator, ExpressionRewriteContext context) {
+        Expression left = rewrite(binaryOperator.left(), context);
+        Expression right = rewrite(binaryOperator.right(), context);
+        if (!TypeCoercionUtils.canHandleTypeCoercion(left.getDataType(), right.getDataType())) {
+            return binaryOperator.withChildren(left, right);
+        }
+        return TypeCoercionUtils.findTightestCommonType(left.getDataType(), right.getDataType())
+                .map(commonType -> {
+                    if (binaryOperator.inputType().acceptsType(commonType) && (
+                            !left.getDataType().equals(commonType) || !right.getDataType().equals(commonType))) {
+                        Expression newLeft = left;
+                        Expression newRight = right;
+                        if (!left.getDataType().equals(commonType)) {
+                            newLeft = new Cast(left, commonType);
```

Comment:
from efficiency perspective, here we will traverse the left child top-down again. Shall we combine this traverse with the traverse in line 61?

Reply:
we could add a memorize supplier to do that just like what plan's output do. I think we could optimize it in a individual PR